### PR TITLE
Add `id` to input props

### LINKED ISSUE #17: RFC: Add `id` to input props
First off, amazing work on this library, it has now fully replaced Formik for me!

Do you think it would be a good idea (perhaps with an opt-out option) to also set the input's `id` to the same value as `name`? I find myself constantly having to add the `id` in order to make it work with `<label for="...">`. On the other hand, setting it automatically may clash with other `id`s the developer wants to use throughout the page.

An alternative solution would be to generate a randomized `id` and expose it as part of the `formState`, or as a third top-level return value from `useFormState()`.

If you think this would be a good addition I'd love to submit a PR for it.

### PR DESCRIPTION
**Pull request by jgierer12**
Fixes #17

### CODE REVIEW

**Comment by jgierer12** [comment]
Note: Some unrelated lines in the readme were auto-formatted by Prettier. I left them in since it's probably better to have a uniform code style throughout the repo but if you'd rather have a clean diff I can revert those lines of course.

**Comment by wsmd** [comment]
Thanks for submitting the PR @jgierer12. I'll review this later this week.

**Comment by wsmd** [comment]
Hey @jgierer12, there's a merge conflict due to recent changes. Could you please take a look?

File: test/userFormState.test.js [hunk 1]
```diff
@@ -113,6 +119,7 @@ describe('input type methods return correct props object', () => {
     expect(input.select('select_name')).toEqual({
       name: 'select_name',
       value: expect.any(String),
+      id: expect.any(String),
```
Comment (wsmd) [inline test/userFormState.test.js]: Once you merge in master, we also need to add this to `selectMultiple`

File: README.md [hunk 2]
```diff
@@ -203,7 +237,7 @@ The following types are currently supported:
 | `<input {...input.time(name: string) />`                    | `{ [name: string]: string }`        |
 | `<select {...input.select(name: string) />`                 | `{ [name: string]: string }`        |
 | `<textarea {...input.textarea(name: string) />`             | `{ [name: string]: string }`        |
-
+| `<label {...input.label(name: string, value: string?)} />`  | N/A                                 |
```
Comment (wsmd) [inline README.md]: I'd suggest making this more descriptive; maybe something like:

```suggestion
| `<label {...input.label(name: string, value?: string)} />`  | N/A – `input.label()` is stateless and thus does not affect the form state |
```

File: README.md [hunk 3]
```diff
@@ -125,6 +125,40 @@ function LoginForm({ onSubmit }) {
...
+    <input {...text('name')} />
+
+    <label {...label('plan', 'free')}>Free Plan</label>
+    <input {...radio('plan', 'free')} />
+
+    <label {...label('plan', 'premium')}>Premium Plan</label>
+    <input {...radio('plan', 'premium')} />
+  </form>
+);
+```
```
Comment (wsmd) [inline README.md]: 💯

**Comment by jgierer12** [comment]
Good thing you added `types/test.tsx`, I would have completely forgotten about adding typings otherwise!

File: README.md (hunk 3 above)
Comment (jgierer12) [reply README.md]: I love how well they line up haha

**Comment by jgierer12** [comment]
By the way, little off-topic tip: You can set
```yml
dist: xenial
```
in `.travis.yml` to avoid these annoying `DeprecationWarning`s, since Ubuntu Xenial has newer Yarn version which doesn't use deprecated node APIs

**Comment by wsmd** [comment]
Oh that would be really cool! I'll have to give that a try. Thanks!

File: src/useFormState.js [hunk 4]
```diff
@@ -71,6 +81,7 @@ export default function useFormState(initialState, options) {
 
     const inputProps = {
       name,
+      id: idGetter(name, ownValue),
```
Comment (wsmd) [inline src/useFormState.js]: Due to #26, we need to call `toString` here

```suggestion
      id: idGetter(name, toString(ownValue)),
```

**Comment by jgierer12** [comment]
```
FAIL  test/useFormState.test.js
  ● Test suite failed to run
    TypeError: Cannot assign to read only property 'Symbol(Symbol.toStringTag)' of object '#<process>'
      at exports.default (node_modules/jest-util/build/create_process_object.js:15:34)
```

I'm not sure this is related to my changes, I am also getting this error on `master` when running `npm run test:all` locally. Could you look into this?

**Comment by wsmd** [comment]
That's odd! Master branch passed 4 days [last time I updated it](https://travis-ci.org/wsmd/react-use-form-state/builds/501248551). I restarted the build on master now and looks like it's failing 😕

I'm investigating.

**Comment by wsmd** [comment]
Alright, updating this branch with master once again should resolve the issue (via https://github.com/wsmd/react-use-form-state/pull/30)

**Comment by wsmd** [comment]
As I was preparing for the 0.8.0 release this morning, I realized that this change can be considered a break change. 

Take the following case for example that may exist today:

```jsx
<div>
  <label htmlFor="form-name" />
  <input id="form-name" {...text('name')} />
</div>
```

The above will be working perfectly fine until we release the new update because we will be overriding the `id` prop on the `input`.

It's arguably a very edge case, where `text()` is called after passing an `id` prop, but, regardless, this is still considered a breaking change.

I'm not ready yet to move to `1.0.0` yet. The only option I see here is to make **this feature available as an opt-in feature**.

```js
const [state, inputs] = useFormState(null, { createIds: true });

// names I'm considering for this option: createIds, useIds, withIds, ids, inputIds
```

Maybe this will change when I get to `1.0.0` to an opt-out. My goal for `1.0.0` is to find a way to avoid passing the name and value twice on both the input and the label as this may be [inconvenient when working with inputs that has long names](https://github.com/wsmd/react-use-form-state/issues/27#issue-416452978).

I'd curious to know what you think.

File: README.md [hunk 5]
```diff
@@ -171,6 +172,40 @@ function LoginForm({ onSubmit }) {
...
+
+An input's generated ID can also be queried with the `id` getter.
+
+```jsx
+const [formState, { id, text }] = useFormState();
+
+return (
+  <>
+    <input {...text('name')} />
+    <p>The input's ID is {id('name')}</p>
```
Comment (wsmd) [inline README.md]: I'm also curious, what's the use case for this?

**Comment by jgierer12** [comment]
Sorry for the late response, been away for a few days. I also think that making it opt-in would be the way to go for the time being. `inputIds` I think would be the most descriptive name. I will open another PR to make this opt-in.

> I'm also curious, what's the use case for this?

If you need the id for something other than labels, e.g. styling, or third-party libs that use the id.

**Comment by jgierer12** [comment]
Ah, my bad I didn't notice you already added this as `withIds`. Still catching up :see_no_evil:

**Pull request merged**